src/starlite_saqlalchemy/repository/sqlalchemy.py

```diff
@@ -136,25 +163,46 @@ async def list(self, *filters: FilterTypes, **kwargs: Any) -> list[ModelT]:
         Returns:
             The list of instances, after filtering applied.
         """
-        for filter_ in filters:
-            match filter_:
-                case LimitOffset(limit, offset):
-                    self._apply_limit_offset_pagination(limit, offset)
-                case BeforeAfter(field_name, before, after):
-                    self._filter_on_datetime_field(field_name, before, after)
-                case CollectionFilter(field_name, values):
-                    self._filter_in_collection(field_name, values)
-                case _:
-                    raise StarliteSaqlalchemyError(f"Unexpected filter: {filter}")
-        self._filter_select_by_kwargs(**kwargs)
+        select_ = self._create_select_for_model()
+        select_ = self._filter_for_list(*filters, select_=select_)
+        select_ = self._filter_select_by_kwargs(select_, **kwargs)
 
         with wrap_sqlalchemy_exception():
-            result = await self._execute()
+            result = await self._execute(select_)
             instances = list(result.scalars())
             for instance in instances:
                 self.session.expunge(instance)
             return instances
 
+    async def list_and_count(
+        self, *filters: FilterTypes, **kwargs: Any
+    ) -> tuple[abc.Sequence[ModelT], int]:
+        """
+
+        Args:
+            *filters: Types for specific filtering operations.
+            **kwargs: Instance attribute value filters.
+
+        Returns:
+            Count of records returned by query, ignoring pagination.
+        """
+        select_ = select(
+            self.model_type,
+            over(sql_func.count(self.model_type.id)),  # type:ignore[attr-defined]
```

Comment:
Nice.  An analytical window function for counting.  This should be fine in the majority of situations.

I'm guessing the other `count` method is there in case we only need to grab a count?

Reply:
Yeh, that was the theory - may as well have both.